### PR TITLE
Fix flash attention output for llama for padded batched inputs

### PR DESCRIPTION
**Pull request by puneeshkhanna**
# What does this PR do?

Fixes # (issue)

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [ ] Did you make sure to update the documentation with your changes?
- [ ] Did you write any new necessary tests?

### CODE REVIEW

**Comment by puneeshkhanna** [comment]
@regisss  - please review and merge.

Triangular masking feature with flash attention is not correctly working for inference right now. We are checking internally.

Without the fix if inputs are left padded which can commonly happen with batched inputs because of different input lens, then with flash attention enabled, we see junk o/p as seen in below command. 

python ../gaudi_spawn.py --use_deepspeed --world_size 8 run_generation.py --model_name_or_path /software/data/llama_inference/Llama-2-70b-hf/ --max_new_tokens 200 --max_input_tokens 50 --bf16 --n_iterations 3 --use_hpu_graphs --use_kv_cache --batch_size 1 --reuse_cache --limit_hpu_graphs --trim_logits --warmup 2 --attn_softmax_bf16 --use_flash_attention

Input/outputs:

input 1: ('DeepSpeed is a machine learning framework',)
output 1: ('DeepSpeed is a machine learning framework for the next 1000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000',)


With the fix in this PR which disables the triangular masking of flash attention for the 1st token generation, output is correct with flash attention enabled.
output 1: ('DeepSpeed is a machine learning framework that enables training of large models with trillions of parameters on a single GPU.\nDeepSpeed is a machine learning framework that enables training of large models with trillions of parameters on a single GPU. It is designed to be used with the PyTorch deep learning framework and can be used to train models with billions of parameters. DeepSpeed is open source and available on GitHub.\nDeepSpeed is a machine learning framework that enables training of large models with trillions of parameters on a single GPU. It is designed to be used with the PyTorch deep learning framework and can be used to train models with billions of parameters. DeepSpeed is open source and available on GitHub. DeepSpeed is a machine learning framework that enables training of large models with trillions of parameters on a single GPU. It is designed to be used with the PyTorch deep learning framework and can be used to train models with billions of parameters. DeepSpeed is open source and',)

Performance wise, results are all fine.

**Review by regisss — COMMENTED** [review]
LGTM!
Please run `make style` also to have the code style check pass.

@mandy-li Would that impact training?

**Comment by puneeshkhanna** [comment]
@regisss  - Done make style.

**Comment by puneeshkhanna** [comment]
@mandy-li  - Did you check ? We need to merge this else with flash attention, we get junk output right now.

**Review by dvarshney-habana — APPROVED** [review]

**Review by regisss — APPROVED** [review]
Let's merge and if there is an issue we'll open a new PR to fix it. It's okay to revert a change in the main branch if needed.

**Comment by mandy-li** [comment]
@regisss , @puneeshkhanna , my intuitive thinking is it won't impact training since FT uses concatenation and trim. But @schoi-habana , please check and test.

**Comment by schoi-habana** [comment]
@regisss @puneeshkhanna @mandy-li There was no impact on eval accuracy and the inference output when using the finetuned model

**Comment by puneeshkhanna** [comment]
Thanks @schoi-habana and @mandy-li for confirmation.
